src/wallet/scriptpubkeyman.cpp

```diff
@@ -0,0 +1,597 @@
+        }
+        batch.ErasePool(index);
+        LogPrintf("keypool index %d removed\n", index);
+        it = setKeyPool->erase(it);
+    }
+}
+
+void ScriptPubKeyMan::MarkUnusedAddresses(const CScript& script)
+{
+    AssertLockHeld(wallet->cs_wallet);
+    // extract addresses and check if they match with an unused keypool key
+    for (const auto& keyid : wallet->GetAffectedKeys(script)) {
+        std::map<CKeyID, int64_t>::const_iterator mi = m_pool_key_to_index.find(keyid);
+        if (mi != m_pool_key_to_index.end()) {
+            LogPrintf("%s: Detected a used keypool key, mark all keypool key up to this key as used\n", __func__);
+            MarkReserveKeysAsUsed(mi->second);
+
+            if (!TopUp()) {
+                LogPrintf("%s: Topping up keypool failed (locked wallet)\n", __func__);
+            }
+        }
+    }
+}
+
+/**
+ * Mark old keypool keys as used,
+ * and generate all new keys
+ */
+bool ScriptPubKeyMan::NewKeyPool()
+{
+    {
+        LOCK(wallet->cs_wallet);
+
+        CWalletDB walletdb(wallet->strWalletFile);
+        // Internal
+        for (const int64_t nIndex : setInternalKeyPool) {
+            walletdb.ErasePool(nIndex);
+        }
+        setInternalKeyPool.clear();
+
+        // External
+        for (const int64_t nIndex : setExternalKeyPool) {
+            walletdb.ErasePool(nIndex);
+        }
+        setExternalKeyPool.clear();
+
+        // Staking
+        for (const int64_t nIndex : setStakingKeyPool) {
+            walletdb.ErasePool(nIndex);
+        }
+        setStakingKeyPool.clear();
+
+        // key -> index.
+        m_pool_key_to_index.clear();
+
+        if (!TopUp()) {
+            return false;
+        }
+        LogPrintf("ScriptPubKeyMan::NewKeyPool rewrote keypool\n");
+    }
+    return true;
+}
+
+/**
+ * Fill the key pool
+ */
+bool ScriptPubKeyMan::TopUp(unsigned int kpSize)
+{
+    if (!CanGenerateKeys()) {
+        return false;
+    }
+    {
+        LOCK(wallet->cs_wallet);
+        if (wallet->IsLocked()) return false;
+
+        // Top up key pool
+        unsigned int nTargetSize;
+        if (kpSize > 0)
+            nTargetSize = kpSize;
+        else
+            nTargetSize = std::max(GetArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 0);
+
+        // Count amount of available keys (internal, external)
+        // make sure the keypool of external and internal keys fits the user selected target (-keypool)
+        int64_t missingExternal = std::max(std::max((int64_t) nTargetSize, (int64_t) 1) - (int64_t)setExternalKeyPool.size(), (int64_t) 0);
+        int64_t missingInternal = std::max(std::max((int64_t) nTargetSize, (int64_t) 1) - (int64_t)setInternalKeyPool.size(), (int64_t) 0);
+        int64_t missingStaking = std::max(std::max((int64_t) nTargetSize, (int64_t) 1) - (int64_t)setStakingKeyPool.size(), (int64_t) 0);
+
+        if (!IsHDEnabled() || !wallet->CanSupportFeature(FEATURE_HD_SPLIT)) {
```

Comment:
same as other places: to check whether `wallet->CanSupportFeature(FEATURE_HD_SPLIT)` seems redundant, when we check also `IsHDEnabled()`. The wallet will either support HD + HD_SPLIT or not, there won't be any intermediate version (with HD enabled and HD_SPLIT not supported).

Reply:
done